src/compiler/checker.ts

```diff
@@ -10739,7 +10739,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             }
         }
         const widened = getWidenedType(addOptionality(type, /*isProperty*/ false, definedInMethod && !definedInConstructor));
-        if (symbol.valueDeclaration && filterType(widened, t => !!(t.flags & ~TypeFlags.Nullable)) === neverType) {
+        if (symbol.valueDeclaration && isInJSFile(symbol.valueDeclaration) && filterType(widened, t => !!(t.flags & ~TypeFlags.Nullable)) === neverType) {

```

Comment:
I think this branch of the code is really meant to be only relevant for JS files but if that's not the case then:
- it would be great to establish what's the appropriate fix for the issue at hand
- it would be great to add some tests relevant to those lines for TS files

Reply:
Isn’t this pattern (expando properties on functions) also valid in TS?  I feel like it is (or should be), since otherwise afaict it’s not really possible to get properties onto a function without a type assertion.

Reply:
yes, it's valid - and this PR makes those nullish cases valid for this pattern as well

Reply:
Then what is the `isInJSFile` check for?  (I couldn’t tell what it’s guarding from the diff, I assumed it was all the logic for expando properties)

Reply:
IIRC JS files have some extra rules when it comes to inferrable things from assignments etc. If I just remove this if statement then tests like [this one](https://github.com/microsoft/TypeScript/blob/e69a0b0e38b380677adb83ff0d20d8b8c32b2d9d/tests/cases/conformance/salsa/inferringClassMembersFromAssignments.ts) change and all of the affected tests are testing JS files. I couldn't relate the observed behavior to any of the behaviors I know from TS files so I concluded that perhaps this was targeting JS files from the beginning (or at least some class of assignments in JS files). So I added this check to ignore this branch in TS files.

Assigned `null`/`undefined` can't widen to anything in TS after all (at least in strict mode) as far as I can tell. So returning `anyType` in TS files here seemed to be just wrong.